llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -147,6 +147,26 @@ getTrackableCopyOps(const MachineInstr &MI, const TargetRegisterInfo &TRI) {
   return {};
 }
 
+/// Returns true if \p MI is a COPY that was eliminated.
+bool trySimplifyCopy(MachineInstr &MI, const PhysRegCopyTracker &CT) {
+  if (!MI.isCopy())
+    return false;
+  auto [DstReg, SrcReg] = MI.getFirst2Regs();
+  if (DstReg.isPhysical() && CT.isCopy(DstReg, SrcReg)) {
+    // Assigning to DstReg a copy of itself. Just remove MI.
+    MI.eraseFromParent();
+    return true;
+  }
+  if (std::optional<Register> EquivVirtSrc;
+      SrcReg.isPhysical() && (EquivVirtSrc = CT.getVirtualCopy(SrcReg))) {
+    // Avoid defining a new VReg if another available one has the same value.
+    MI.getMF()->getRegInfo().replaceRegWith(DstReg, *EquivVirtSrc);
+    MI.eraseFromParent();
+    return true;
```

Comment:
Can we now guarantee we never have more than one VirtReg equivalent to a particular PhysReg?

Reply:
Could we be mixing reg classes here and get in trouble similar to the grief we've seen in the singleton regclass for SELECT?

Reply:
> Can we now guarantee we never have more than one VirtReg equivalent to a particular PhysReg

I think we can actually, I just wanted to keep the tracker generic if we later extend to an interblock analysis.